added first_differences method

nbokulich opened the pull request:
fixes #58 

This computes "first differences", or the difference/distance between sequential measurements taken from the same individual. This simply computes the differences, which can then be used as input for LME or other methods.

nbokulich committed:
added first_differences method

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -183,12 +183,15 @@ def _get_pairwise_differences(df, pairs, category, individual_id_column,
         pre_value = float(df[category][pre_idx])
         post_value = float(df[category][post_idx])
         paired_difference = post_value - pre_value
+        print(paired_difference)
```
Can you remove this print statement?

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -28,14 +30,20 @@
 base_parameters = {
     **shared_parameters,
-    'state_column': Str,
+    **{k: v for k, v in miscellaneous_parameters.items()
```
This approach may be clearer and avoids a loop that's only looking for a particular item:

```python
'state_column': miscellaneous_parameters['state_column']
```

If you agree with this change, can you make it here and in the other places that are looping over `miscellaneous_parameters` and `miscellaneous_parameter_descriptions`?

nbokulich replied:
:man_facepalming: ha! thanks @jairideout ! I think I needed some more 🍵 when I wrote that...

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +288,33 @@ def test_nmit_missing_table_ids(self):
+        self.assertTrue(obs.sort_index().equals(exp.sort_index()))
+
+    def test_first_differences_taxa(self):
+        first_differences(
```
This test isn't asserting the results are correct.

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +288,33 @@ def test_nmit_missing_table_ids(self):
+            metadata=qiime2.Metadata(md), state_column='Time',
+            individual_id_column='ind', replicate_handling='drop',
+            distance_matrix=dm)
+        self.assertTrue(obs.equals(exp))
```
You can compare Series and DataFrame objects using some pandas testing utilities:

```python
import pandas.util.testing as pdt

pdt.assert_series_equal(obs, exp)
# OR:
pdt.assert_frame_equal(obs, exp)
```

This is a more common approach to asserting that pandas objects are equal, and it's handy because it'll tell you what isn't the same between the two objects. We use this strategy in several qiime2 plugins (e.g. q2-diversity), so there's some other examples floating around if it's helpful to you.

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -48,7 +56,19 @@
+    'replicate_handling': (
+        'Choose how replicate samples are handled. If replicates are '
+        'detected, "error" causes method to fail; "drop" will discard all '
+        'subject IDs with replicate samples at either state_1 or state_2; '
```
I don't think there are `state_1` or `state_2` parameters for `first_differences`.

nbokulich replied:
thanks! You are correct — that info is out of date.

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -48,7 +56,19 @@
+        'replicates.')
 }
 
 shared_parameter_descriptions = {
```
I can't comment on the exact line, but the parameter description for `metadata` (below) references `group_column`, which doesn't exist for `first_differences`.

jairideout commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -192,3 +196,58 @@ def nmit(table: pd.DataFrame, metadata: qiime2.Metadata,
+
+def first_differences(metadata: qiime2.Metadata, state_column: str,
+                      individual_id_column: str, metric: str='Distance',
+                      replicate_handling: str='error',
```
I'm getting a misformatted error message when replicates are detected (default behavior):

```
Plugin error from longitudinal:

  Replicate values for individual {0} at state {1}. Remove replicate
  values from input files or set replicate_handling parameter to select
  how replicates are handled.

Debug info has been saved to /tmp/qiime2-q2cli-err-pd24umjc.log
```

The values aren't being interpolated, and perhaps it might be clearer if it read:

*Detected replicate samples for individual ...*

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +237,39 @@
+                'metric': Str},
+    outputs=[('first_differences', SampleData[AlphaDiversity])],
+    input_descriptions={
+        'table': 'Feature table to optionally use for paired comparisons.',
```
Are these "paired comparisons"? I can't tell if this description is accurate or a copy-n-paste from `pairwise-differences`.

nbokulich replied:
thanks! technically this method is chaining together a bunch of paired comparisons, but I have modified to make this clearer.

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -48,7 +56,19 @@
+
+miscellaneous_parameter_descriptions = {
+    'state_column': ('Metadata column containing state (e.g., Time) '
+                     'across which samples are paired.'),
```
Are samples being paired in `first_differences`? I can't tell if this description is accurate or a copy-n-paste from `pairwise-differences`.

nbokulich replied:
Yes, thanks for checking — do you think this needs to be clarified further here?

jairideout replied:
The word "paired" threw me off because I initially thought it meant that there were only two states (e.g. "before" and "after"). I see now that these samples are being "paired" sequentially. I don't have a suggestion to make it clearer, fine to leave as-is unless my thought process above helps prompt some text changes :)

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +237,39 @@
+    output_descriptions={'first_differences': 'Series of first differences.'},
+    name='First difference computation between sequential states',
+    description=(
+        'Calculates first differences in "metric" between sequential states '
```
`state_column` allows both numeric and categorical metadata columns, but there isn't a guaranteed ordering to metadata columns in qiime2. How does this method order states such that they are sequential? If this column was required to be numeric, I guess the values could be sorted to form a sequence, but what happens if categorical data is supplied? I'm not sure if this is a problem or not, but wanted to bring it up in case it's a "gotcha".

nbokulich replied:
thanks @jairideout ! This is a very good point — I think that for this method (and LME and volatility, which would be prone to the same issue) we should require state_column to be numeric... and I've added a validator (pass the data through the same validator used for the above issue). I think that's the best solution here, but am open to other ideas. For other methods, I do not want to enforce this, e.g., for paired methods.

jairideout replied:
:+1: that sounds like a good approach! With this new validator, would the other methods (LME and volatility) automatically get the fix in this PR, or are you thinking of a separate PR to update those methods? I'm fine either way, just want to know what to be on the lookout for.

nbokulich replied:
Yep, since it's a small change I slipped that validator into LME and volatility in this PR.

jairideout replied:
Sounds good!

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +237,39 @@
+    parameter_descriptions={
+        **miscellaneous_parameter_descriptions,
+        **shared_parameter_descriptions,
+        'metric': ('Numerical metadata or artifact column to test. If '
```
When `metric` is non-numeric can you make the error message clearer?

```
Plugin error from longitudinal:

  could not convert string to float: 'n'

Debug info has been saved to /tmp/qiime2-q2cli-err-dofhig1q.log
```

nbokulich replied:
thanks! I've added a new validator to test this during metadata merging

jairideout replied:
I'm still getting the same error message:

```
$ qiime longitudinal first-differences --m-metadata-file ecam-sample-metadata.tsv --p-state-column month --p-individual-id-column studyid --o-first-differences out.qza --p-replicate-handling drop --p-metric antiexposedall
Plugin error from longitudinal:

  could not convert string to float: 'n'

Debug info has been saved to /tmp/qiime2-q2cli-err-z3gp87xf.log
```

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +288,33 @@ def test_nmit_missing_table_ids(self):
             nmit(table=self.table_taxa_fp, metadata=md,
                  individual_id_column='studyid')
 
+    def test_first_differences(self):
```
General comment: these tests are pretty light, can you add some more that hit border cases and error cases? This list isn't exhaustive, but here's some examples:

- What happens when there's no states, a single state, only two states?
- What happens when there's no individuals, or a single individual?
- What happens when the `table` is empty? Single-sample only?
- What happens if the `distance_matrix` is single-sample only?
- What happens when there are no differences across states?
- Can you test different types of replicate handling other than `drop`?
- What happens if there are no replicates? 
- What happens when an ID in the feature table isn't found in the metadata? How about for a distance matrix?
- Are extra IDs in the metadata ignored when they're not found in the table or distance matrix?
- What if the same column name is supplied for `individual_id_column` and `state_column`? (should probably error).
- What happens when the same column is used for `metric` and either `individual_id_column` or `state_column`? If `table` is provided, that behavior is probably okay (but worth having a unit test for). If only `metadata` is provided, an error is probably appropriate.
- There's only one test data set being used in the unit tests. Can you add another valid test data set so that we're not overfitting the tests to a single data set?
- What happens when `metric` is non-numeric?

nbokulich replied:
Thanks @jairideout ! I have added a whole bunch of tests. Many of these issues (e.g., column validation) are checked by utility functions that already existed, but I have added tests for some of these anyway. I have also added several more validator functions in response to your points — you bring up a number of very good cases!

> If table is provided, that behavior is probably okay (but worth having a unit test for)

Why do you think it would be okay for metric to match other columns if the metric is in a table? I can make workarounds for that case, but for now I've added a validator to make sure *no columns* match because as far as I can think this may lead to unintended consequences.

> Can you add another valid test data set

I have one real data set and one toy data set already implemented. What kind of additional datasets would you want to see added (real or toy)?

jairideout replied:
> Why do you think it would be okay for metric to match other columns if the metric is in a table? I can make workarounds for that case, but for now I've added a validator to make sure no columns match because as far as I can think this may lead to unintended consequences.

I was thinking it'd be technically acceptable to supply a `metric` in the feature table which could overlap with the same column name in `metadata` used for `individual_id_column` or `state_column`. We know it isn't the same column being supplied because they're coming from different files (`table` vs `metadata`). For example, imagine your `metric` in the feature table is called `foo`, and you also happen to have an `individual_id_column` called `foo`. Those `foo` names refer to columns in different files so it *could* be a valid operation, but also a very odd border case.

I think this is such a rare/odd border case, and the behavior wouldn't be clear in that situation unless the user carefully read the method's docs. The validator behavior you described is better in the long run, let's go with that!

> I have one real data set and one toy data set already implemented. What kind of additional datasets would you want to see added (real or toy)?

Apologies, I thought there was only one "valid" data set being tested, I missed the other one. I think with the addition of the data sets necessary to test those border cases I listed above, we should be GTG!

jairideout replied:
Thanks for adding these test cases! :+1: 

Can you also add a test case where the sequential states aren't in any kind of order in the metadata file (i.e. to test that they are sorted correctly within the method)? It'd also be good to have a test case where the order of sample IDs in the metadata file don't match the order of sample IDs in the table or distance matrix. Apologies if you already have these test cases, from a quick look I didn't see it.

jairideout commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -192,3 +196,58 @@ def nmit(table: pd.DataFrame, metadata: qiime2.Metadata,
     return _dist
+
+
+def first_differences(metadata: qiime2.Metadata, state_column: str,
```
I played around with `first_differences` locally and have some suggestions for tweaking the interface. There's a few mutually-exclusive inputs to this method, so it may be worth breaking apart into separate methods that operate on metadata-only, metadata + table, and metadata + distance matrix.

If you want to keep this a single method, these changes to the interface may make it clearer (and avoid some undefined edge cases):

- Make `metric` have a default value of `None`, and error if `metric is not None` and a `distance_matrix` was provided. The current default of `Distance` is a little ambiguous because there could be a metadata column called `Distance`, or (technically) it could be a feature ID in `table`. By having a default value of `None` we're not making a statement about the metric, and can detect when the user is providing invalid input (i.e. a `distance_matrix` and a `metric`, which is disallowed).

- If `table` and `distance_matrix` are both provided, error.

- If `metric` is not provided and either `table` + `metadata` or `metadata`-only is provided, error.

- If `metric` is provided with `table` + `metadata`, the metric is looked up in `table` and errors otherwise, instead of falling back to finding the column in `metadata`.

nbokulich replied:
Thanks for the feedback @jairideout — I think you are correct, it makes most sense to split this into multiple methods.

However, I would argue that splitting into 2 methods makes the most sense; one method that operates on a distance matrix and one that operates on metadata (and accepts a feature table as optional input). For the latter case, a user would always need to define `metric` whether or not a table is input, and we could always raise an error if `metric` is not a column in the table. How does that sound?

jairideout replied:
:+1: That sounds perfect!

jairideout reviewed (changes requested):
Thanks for implementing this new method @nbokulich! I didn't do a detailed review of the code yet. I focused this first-pass review on blackbox testing by pulling the PR down locally and testing it out as a user would. I have some interface changes to suggest (see comments inline). Another larger request is to add more unit tests -- I described some additional cases to test in the review comments.

Can you take a look at the review and let me know what you think? I'll hold off on doing a more in-depth review until we can finalize the interface and behavior. Thanks!

nbokulich committed:
split into first_differences and first_distances methods

nbokulich committed:
first_distances tests

nbokulich committed:
fixed replicates detected error

jairideout commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -184,11 +196,51 @@ def nmit(table: pd.DataFrame, metadata: qiime2.Metadata,
+
+
+def first_differences(metadata: qiime2.Metadata, state_column: str,
+                      individual_id_column: str, metric: str='Distance',
```
Can you make `metric` required here?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -184,11 +196,51 @@ def nmit(table: pd.DataFrame, metadata: qiime2.Metadata,
+                    state_column: str, individual_id_column: str,
+                    replicate_handling: str='error') -> pd.Series:
+
+    # find metric in metadata or derive from table and merge into metadata
```
outdated comment

nbokulich replied:
👍

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -650,9 +674,71 @@ def _nmit(table, sample_md, individual_id_column, corr_method="kendall",
+
+    # create dummy group column in metadata so we can use downstream functions
+    # that split metadata by groups without actually bothering to do so.
+    group_column = "dummy_group"
```
This would overwrite user data if there happens to be a column called `dummy_group` already in the metadata. Is there another way to handle this without adding a dummy column, or somehow guaranteeing a unique name that won't ever be found in the user's metadata?

nbokulich replied:
you are right... I cannot think of an easy workaround since we need to create this dummy column in the same dataframe. It should not be an issue unless if "dummy_group" is the name of `state_column`, `individual_id_column`, or `metric`.

The best solution I can think of right now is to rename this to: `"a_name_that_no_user_of_qiime2_could_conceivably_ever_use"`

👍 👎 ❓

jairideout replied:
What about calling `uuid.uuid4()` until you find a UUID that isn't a column name in the table? It's hacky but will never collide with a user's columns. We're taking this approach in `Metadata.ids()`, you could reuse that implementation here:

https://github.com/qiime2/qiime2/blob/master/qiime2/metadata.py#L294

nbokulich replied:
thanks @jairideout ! I love this solution!!!

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -650,9 +674,71 @@ def _nmit(table, sample_md, individual_id_column, corr_method="kendall",
+            state_values=[states[s], states[s + 1]],
+            replicate_handling=replicate_handling)
+        # compute distance between pairs
+        if metric == 'Distance':
```
I don't think this check is safe, what if the user runs `first_differences` with a `metric` that's named `'Distance'` (i.e. in their metadata file or table)? In general you'll want to avoid using sentinel values that could collide with user-defined names, otherwise the code can't tell whether the user provided the value or if it is a sentinel. This comment also relates to using `'dummy_group'` above for a sentinel.

nbokulich replied:
thanks! Right as always! I have changed this to the much more effective and appropriate `if distance_matrix is not None`

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+                distance_matrix=dm_single_sample, metadata=qiime2.Metadata(md),
+                state_column='Time', individual_id_column='ind',
+                replicate_handling='drop')
+            print(obs)
```
remove print statement

nbokulich replied:
😨

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+
+    def test_first_distances_single_sample(self):
+        with self.assertRaisesRegex(RuntimeError, "Output is empty"):
+            obs = first_distances(
```
This is a minor stylistic comment, up to you whether to address or leave as-is:

You don't need to store the output of a method in a variable that's expected to error, so it can be simplified to:

```python
with self.assertRaisesRegex(RuntimeError, "Output is empty"):
    first_distances(...)
```

nbokulich replied:
👍

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -59,16 +60,29 @@ def _validate_input_values(df, individual_id_column, group_column,
+    if state_column is not None:
+        states = df[state_column].unique()
+        if len(states) < 2:
+            raise ValueError(("state_column must contain at least two unique "
```
This is a minor stylistic comment, up to you whether to address or leave as-is (it applies to a few places in your PR):

You don't need to wrap string literals in parentheses if they're already contained within parentheses or any sort of braces/brackets (e.g. `[ ]`, `{ }`).

```python
raise ValueError("state_column must contain at least two unique "
                 "values. Values in {0}: {1}".format(state_column, states))
```

nbokulich replied:
👍 thanks! (I fixed all cases, so you may note some utilities not related to `first-differences` appear in the next push)

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -650,9 +674,71 @@ def _nmit(table, sample_md, individual_id_column, corr_method="kendall",
+    pairs_summary = pd.DataFrame()
+    errors = []
+    states = metadata[state_column].unique()
+    # iterate over range of sorted states in order to compare sequential states
```
Where is the sorting happening? `pd.Series.unique()` explicitly doesn't guarantee an output order:

https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.unique.html

nbokulich replied:
Thanks! `pd.Series.unique()` has always seemed to sort by default so thank you for revealing this. I have added sorting here and one other spot where order matters.

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +232,68 @@
+    parameters={**miscellaneous_parameters,
+                **shared_parameters,
+                'metric': Str},
+    outputs=[('first_differences', SampleData[AlphaDiversity])],
```
Sorry I didn't notice this on the first review: these methods are producing `SampleData[AlphaDiversity]` artifacts. Conceptually, are the outputs of `first_differences` and `first_distances` alpha diversity measurements, or is this just a convenient way to store per-sample values in (i.e. because they can be viewed as metadata)?

If it's the latter case, it probably makes sense to define a new semantic type `SampleData[Something]` that captures what these values are, and that type could have `pd.Series` and `Metadata` transformers like `SampleData[AlphaDiversity]` does.

nbokulich replied:
I have been waiting to have this conversation...

It is a bit of cases A and B. The underlying data will often be alpha diversity data and I want the outputs to be useable in many of the same ways as alpha diversity (not necessarily just as metadata). But many of the data types might not be alpha diversity data (e.g., if a taxon is input as `metric`) so I think you are right I just need to define a new semantic type.

`SampleData[FirstDifferences]`?

nbokulich replied:
I have defined the `SampleData[FirstDifferences]` type now

jairideout replied:
Awesome, @ebolyen and I :+1: the name! Thanks for adding that.

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +232,68 @@
+        **miscellaneous_parameter_descriptions,
+        **shared_parameter_descriptions,
+        'metric': ('Numerical metadata or artifact column to test. If '
+                   'computing first differences on a distance matrix, do not '
```
outdated docs now that distance matrix support is in its own method

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -221,3 +232,68 @@
+    description=(
+        'Calculates first distances between sequential states for samples '
+        'collected from individual subjects sampled repeatedly at two or more '
+        'states. This method is similar to the first differences method, '
```
may be clearer to quote "first differences" or refer to it as `first_differences`

nbokulich replied:
👍

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+            metric='Value', replicate_handling='random')
+        # The first diff of individual 2 is subject to random rep handling
+        mystery_number = obs.iloc[1]
+        self.assertTrue(mystery_number in [0.04999999999999999, 0.06])
```
This check could fail on different platforms/architectures due to floating point error. Can you use [`self.assertAlmostEqual()`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual) instead of exact equality?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_taxa(self):
+        first_differences(
```
Can you assert the results are correct here?

nbokulich replied:
👍 I've added the expected results as a tsv

jairideout commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+            replicate_handling='drop')
+        pdt.assert_series_equal(obs, exp)
+
+    # just check that it works on real data
```
Can you assert the results somehow in this test?

nbokulich replied:
👍 I've added the expected results as a tsv

jairideout reviewed (changes requested):
Thanks for the updates @nbokulich, the new tests are great! I had some more review comments after taking a second pass through.

nbokulich committed:
define FirstDifferences semantic type

nbokulich committed:
added _validate_ attributes to FirstDifferences type

nbokulich committed:
resolved merge conflicts

ebolyen committed:
MAINT: add missing annotations (#63)

jairideout commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -184,11 +196,51 @@ def nmit(table: pd.DataFrame, metadata: qiime2.Metadata,
+        metadata, individual_id_column, None, state_column, None)
+
+    return _first_differences(
+        metadata, state_column, individual_id_column, metric='Distance',
```
Can you pass `metric=None` here? `metric` is ignored when a `distance_matrix` is provided, so passing a null value might avoid confusion about `metric` actually being used to lookup values somewhere.

nbokulich replied:
thanks for catching @jairideout !

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -226,3 +306,66 @@
+    parameter_descriptions={
+        **miscellaneous_parameter_descriptions,
+        **shared_parameter_descriptions,
+        'metric': 'Numerical metadata or artifact column to test.',
```
I think this comment got lost in the last round of review. Can you make this error message a little clearer when a non-numeric metric is supplied?

```
$ qiime longitudinal first-differences --m-metadata-file ecam-sample-metadata.tsv --p-state-column month --p-individual-id-column studyid --o-first-differences out.qza --p-replicate-handling drop --p-metric antiexposedall
Plugin error from longitudinal:

  could not convert string to float: 'n'

Debug info has been saved to /tmp/qiime2-q2cli-err-z3gp87xf.log
```

nbokulich replied:
sorry about that! yes, I think it got buried in the split review last time.

I added a new unit test to catch this and found the spot where this was missing (I previously only raised the error when merging metric from a feature table).

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+
+
+class FirstDifferencesFormat(model.TextFileFormat):
+    def _check_n_records(self, n_lines=None):
```
If `n_lines=None` an error will be raised. Can you have it iterate through the entire file when `n_lines=None`? That'll help when the user requests `max` validation (see related comment below).

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+class FirstDifferencesFormat(model.TextFileFormat):
+    def _check_n_records(self, n_lines=None):
+        with self.open() as fh:
+            line = fh.readline()
```
Can you validate the header is what's expected? i.e. `#SampleID\tDifference` or `#SampleID\tDistance`?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+            for line, _ in zip(fh, range(n_lines)):
+                cells = line.strip().split('\t')
+                if len(cells) != 2:
+                    raise ValidationError('File is not FirstDifferencesFormat')
```
Can you describe why the file is invalid in the error message? That'll help users debug their data. The framework will wrap your error message to tell the user which format is in question, so you don't need to include that boilerplate here. Something like *"Expected two columns, found %d"* could work.

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+                    raise ValidationError('File is not FirstDifferencesFormat')
+
+    def _validate_(self, level):
+        record_count_map = {'min': 5, 'max': 1000}
```
Can you supply `max: None` instead of `1000` to have it validate the entire file?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+            line = fh.readline()
+            for line, _ in zip(fh, range(n_lines)):
+                cells = line.strip().split('\t')
+                if len(cells) != 2:
```
Can you also assert that the second cell of each row is a number?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
 )
 
 
+# FirstDifferencesFormat is a near replica of AlphaDiversityFormat in q2_types
```
Can you move all of the semantic type and format code into a new file, e.g. `_format.py` or `_type.py`? That'll help keep `plugin_setup.py` clean and matches some of the conventions we've been using elsewhere.

nbokulich replied:
Please @jairideout don't send me to circular import hell! 🌀🔥😱🔥🌀 maybe I could fix this in a future PR? (you could raise an issue to make sure I do it!) at this point I'm worried about getting this PR into the release...

jairideout commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -29,14 +35,84 @@
+
+
+@plugin.register_transformer
+def _5(ff: FirstDifferencesFormat) -> (pd.Series):
```
For this transformer and the Metadata transformer below, can you cast the column to float so that the differences aren't a string dtype in the object?

nbokulich replied:
👍

jairideout commented on q2_longitudinal/_utilities.py:
```diff
@@ -683,9 +707,71 @@ def _nmit(table, sample_md, individual_id_column, corr_method="kendall",
                          table_ids.difference(metadata_ids))
+
+
+def _validate_is_numeric_column(metadata, metric):
```
Can you have this utility also cast the column into a float dtype? I think it's possible for a user to supply a column that's composed of strings representing numbers -- I'm not sure what downstream code would do in that case with the "string numbers" (it might sort things lexicographically, which would be unexpected). Maybe worth a unit test too, something with states `1` and `10`?

nbokulich replied:
check out the new changes to `_add_metric_to_metadata`. If the metric is in metadata, it will already be cast to float by `_load_metadata`. The one metric not covered previously (I think) would have been metadata coming from feature tables.

See also new test `test_first_distances_numeric_values_represented_as_strings` to cover this.

jairideout reviewed (changes requested):
Final round of review is done, thanks for all of these changes @nbokulich! I had some more comments -- I think most are easy to address.

nbokulich committed:
expanded semantic type tests

nbokulich committed:
flaked

jairideout committed:
Merge branch 'issue-58' of https://github.com/nbokulich/q2-intervention into nbokulich-issue-58

jairideout committed:
SQUASH: misc updates described in detail below

Took a final pass review and made the following changes:

- Moved type, format, and transformer code out of plugin_setup.py and into their respective modules `_type.py`, `_format.py`, `_transformer.py`. Also moved related tests out of test_longitudinal.py and into their respective test modules `test_type.py`, `test_format.py`, `test_transformer.py`.

- Improvements to the FirstDifferencesFormat validator, including more descriptive error messages with less boilerplate text, correct line number reporting, and detecting absence of data records.

- Added more unit tests for FirstDifferencesFormat (this also improved its coverage report).

- Transformers now error if data is non-numeric when being read from file.

- Transformer unit tests are no longer as circular as before in their declaration of expected values.

- Removed `longitudinalTestPluginBase` in favor of the framework's `TestPluginBase` (the former was a reimplementation of the latter).

- Added `__main__` guard to all test modules so that individual test files are executable with `python path/to/my/test_file.py`

- Reordered and grouped imports in all modules to match the following convention: 1) standard library imports, 2) third-party imports, 3) current package imports

- Made some of the `from foo import bar` style imports into `import foo` to keep things namespaced. There's many more imports like that to update, but that's for another day :)
